### PR TITLE
Bug 1147584 - Don't remove unlikely <a> tags, and replace <a> tags with ...

### PR DESCRIPTION
**Pull request by leibovic**
...their text content if they won't be useful links

r? @gijsk 

This alters the results of some testcases, but I think for the better (removes useless <a> tags).

### CODE REVIEW

**Comment by leibovic** [comment]
I should note, the reason that we were removing these <a> elements in the testcase [1] is that they have a class="popup".

[1] http://www.vcu.edu/engweb/webtexts/bartleby/

**Comment by n1k0** [comment]
FWIW, r=me

File: Readability.js [hunk 1]
```diff
@@ -524,7 +529,8 @@ Readability.prototype = {
         if (stripUnlikelyCandidates) {
           if (this.REGEXPS.unlikelyCandidates.test(matchString) &&
               !this.REGEXPS.okMaybeItsACandidate.test(matchString) &&
-              node.tagName !== "BODY") {
+              node.tagName !== "BODY" &&
+              node.tagName !== "A") {
```
Comment (gijsk) [inline Readability.js]: Is this addition really necessary? It seems unlikely to really impact candidate selection (which is mostly paragraph-related anyway, I think?).

File: Readability.js [hunk 2]
```diff
@@ -181,20 +181,25 @@ Readability.prototype = {
...
-          elem.setAttribute(propName, toAbsoluteURI(relativeURI));
-      });
-    }
-
-     // Fix links.
-    convertRelativeURIs.call(this, "a", "href");
+    var links = articleContent.getElementsByTagName("a");
+    this._forEachNode(links, function(link) {
+      var href = link.getAttribute("href");
+      if (!href || href.indexOf("javascript:") === 0) {
```
Comment (gijsk) [inline Readability.js]: This is going to break two things:

* JS-based `window.open` calls with appropriate URIs
* Anchor refs inside the document... but those were already broken, I think? Maybe?

File: Readability.js (hunk 1 above)
Comment (gijsk) [inline Readability.js]: Oh, hm, I see you commented on this matter earlier. Nevermind me!

File: Readability.js [hunk 3]
```diff
@@ -197,20 +197,29 @@ Readability.prototype = {
...
+    var links = articleContent.getElementsByTagName("a");
+    this._forEachNode(links, function(link) {
+      var href = link.getAttribute("href");
+      if (href) {
+        // Replace links with javascript: URIs with text content, since
+        // they won't work after scripts have been removed from the page.
+        if (href.indexOf("javascript:") === 0) {
+          var text = new Text();
+          text.textContent = link.textContent;
+          link.parentNode.replaceChild(text, link);
```
Comment (leibovic) [inline Readability.js]: I updated this to be a bit less aggressive and only replace links with javascript: URLs with their text content. However, I was running into a jsdom error with replaceChild when testing locally:

  1) Test pages clean-links jsdom "before all" hook:
     TypeError: Object [object Object] has no method 'dispatchEvent'
      at core.Node.insertBefore (/Users/leibovic/code/readability/node_modules/jsdom/lib/jsdom/level2/events.js:332:14)
      at core.Node.replaceChild (/Users/leibovic/code/readability/node_modules/jsdom/lib/jsdom/level1/core.js:707:10)
      at Object.<anonymous> (/Users/leibovic/code/readability/Readability.js:209:27)
      at NodeList.forEach (native)
      at Object.Readability._forEachNode (/Users/leibovic/code/readability/Readability.js:149:36)
      at Object.Readability._fixRelativeUris (/Users/leibovic/code/readability/Readability.js:201:10)
      at Object.Readability._postProcessContent (/Users/leibovic/code/readability/Readability.js:134:10)
      at Object.Readability.parse (/Users/leibovic/code/readability/Readability.js:1721:10)
      at /Users/leibovic/code/readability/test/test-readability.js:110:34
      at Context.<anonymous> (/Users/leibovic/code/readability/test/test-readability.js:17:16)
      at callFn (/Users/leibovic/code/readability/node_modules/mocha/lib/runnable.js:266:21)
      at Hook.Runnable.run (/Users/leibovic/code/readability/node_modules/mocha/lib/runnable.js:259:7)
      at next (/Users/leibovic/code/readability/node_modules/mocha/lib/runner.js:268:10)
      at Object._onImmediate (/Users/leibovic/code/readability/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:345:15)

I'm not sure what's going on here. @n1k0 @gijsk ideas?

Comment (n1k0) [inline Readability.js]: This seems to be jsdom-only… dunno what is happening there.

Also, while generating readable content using the patch (using JSDOMParser), things like:

```html
<div>
  Lorem <a href="javascript:void(0)">ipsum</a> dolor sit amet
</div>
```

(notice we're using a `div`, not a `p`)

Gives

```html
<div>
    <p style="display: inline;" class="readability-styled">Lorem</p>ipsum
    <p style="display: inline;" class="readability-styled">dolor sit amet Ut enim ad minim veniam, quis nostrud exercitation ullamco
        laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor
        in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
        pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa
        qui officia deserunt mollit anim id est laborum.</p>
</div>
```

Comment (gijsk) [inline Readability.js]: This is expected. I tried to reduce this paragraph-creation logic (it's in the "experimental" block in Readability.js), but it caused a lot of test failures because things get weighted by how many paragraphs are in them, and obviously this is skewing the numbers. :-)

Comment (gijsk) [inline Readability.js]: Oh, and in relation to the jsdom error... I expect that what is going on is that jsdom incorrectly assumes that the second argument to `replaceChild` has to be an element. (you can't `dispatchEvent` on a text node)
This assumption is wrong, and our code is exposing that problem here... Sadly, I don't really know a good way around this besides using `insertBefore` followed by `removeChild`... except that we don't currently implement `insertBefore` in JSDOMParser...

Comment (gijsk) [inline Readability.js]: I have a branch that implements `insertBefore` now, and then I read the error log and realized that this wasn't going to help.

Disabling `MutationEvents` helps but then I just get a new error:

  1) Test pages clean-links jsdom "before all" hook:
     TypeError: Cannot call method 'replace' of undefined
      at escapeString (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:26:10)
      at Serializer._serializeTextNode (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:169:56)
      at Serializer._serializeChildNodes (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:78:22)
      at Serializer._serializeElement (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:118:14)
      at Serializer._serializeChildNodes (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:75:22)
      at Serializer._serializeElement (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:118:14)
      at Serializer._serializeChildNodes (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:75:22)
      at Serializer._serializeElement (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:118:14)
      at Serializer._serializeChildNodes (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:75:22)
      at Serializer.serialize (/Users/gkruitbosch/dev/readability/node_modules/jsdom/node_modules/parse5/lib/serialization/serializer.js:60:10)
      at exports.domToHtml (/Users/gkruitbosch/dev/readability/node_modules/jsdom/lib/jsdom/browser/domtohtml.js:16:20)
      at innerHTML (/Users/gkruitbosch/dev/readability/node_modules/jsdom/lib/jsdom/level1/core.js:1409:12)
      at Object.Readability.parse (/Users/gkruitbosch/dev/readability/Readability.js:1745:37)
      at /Users/gkruitbosch/dev/readability/test/test-readability.js:111:34
      at Context.<anonymous> (/Users/gkruitbosch/dev/readability/test/test-readability.js:17:16)
      at callFn (/Users/gkruitbosch/dev/readability/node_modules/mocha/lib/runnable.js:266:21)
      at Hook.Runnable.run (/Users/gkruitbosch/dev/readability/node_modules/mocha/lib/runnable.js:259:7)
      at next (/Users/gkruitbosch/dev/readability/node_modules/mocha/lib/runner.js:268:10)
      at Object._onImmediate (/Users/gkruitbosch/dev/readability/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:345:15)

**Comment by gijsk** [comment]
Closing this in favour of #135, which is the same but fixes the tests and implements `createTextNode` .
